core/src/main/java/org/apache/iceberg/encryption/Ciphers.java

```diff
@@ -0,0 +1,125 @@
+      this.aesKey = new SecretKeySpec(keyBytes, "AES");
+
+      try {
+        this.cipher = Cipher.getInstance("AES/GCM/NoPadding");
+      } catch (GeneralSecurityException e) {
+        throw new RuntimeException("Failed to create GCM cipher", e);
+      }
+
+      this.randomGenerator = new SecureRandom();
+    }
+
+    public byte[] encrypt(byte[] plainText, byte[] aad) {
+      byte[] nonce = new byte[NONCE_LENGTH];
+      randomGenerator.nextBytes(nonce);
+      int cipherTextLength = NONCE_LENGTH + plainText.length + GCM_TAG_LENGTH;
+      byte[] cipherText = new byte[cipherTextLength];
+
+      try {
+        GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, nonce);
+        cipher.init(Cipher.ENCRYPT_MODE, aesKey, spec);
+        if (null != aad) {
+          cipher.updateAAD(aad);
+        }
+        cipher.doFinal(plainText, 0, plainText.length, cipherText, NONCE_LENGTH);
+      } catch (GeneralSecurityException e) {
+        throw new RuntimeException("Failed to encrypt", e);
+      }
+
+      // Add the nonce
+      System.arraycopy(nonce, 0, cipherText, 0, NONCE_LENGTH);
+
+      return cipherText;
+    }
+  }
+
+  public static class AesGcmDecryptor {
+    private final SecretKeySpec aesKey;
+    private final Cipher cipher;
+
+    public AesGcmDecryptor(byte[] keyBytes) {
+      int keyLength = keyBytes.length;
+      Preconditions.checkArgument((keyLength == 16 || keyLength == 24 || keyLength == 32),
+          "Cannot use a key of length " + keyLength + " because AES only allows 16, 24 or 32 bytes");
+      this.aesKey = new SecretKeySpec(keyBytes, "AES");
+
+      try {
+        this.cipher = Cipher.getInstance("AES/GCM/NoPadding");
+      } catch (GeneralSecurityException e) {
+        throw new RuntimeException("Failed to create GCM cipher", e);
+      }
+    }
+
+    public byte[] decrypt(byte[] ciphertext, byte[] aad)  {
+      int plainTextLength = ciphertext.length - GCM_TAG_LENGTH - NONCE_LENGTH;
+      Preconditions.checkState(plainTextLength >= 1,
+          "Cannot decrypt cipher text of length " + ciphertext.length +
+          " because text must longer than GCM_TAG_LENGTH + NONCE_LENGTH bytes. Text may not be encrypted" +
+          " with AES GCM cipher");
+
+      // Get the nonce from ciphertext
+      byte[] nonce = new byte[NONCE_LENGTH];
+      System.arraycopy(ciphertext, 0, nonce, 0, NONCE_LENGTH);
+
+      byte[] plainText = new byte[plainTextLength];
+      int inputLength = ciphertext.length - NONCE_LENGTH;
+      try {
+        GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, nonce);
+        cipher.init(Cipher.DECRYPT_MODE, aesKey, spec);
+        if (null != aad) {
+          cipher.updateAAD(aad);
+        }
+        cipher.doFinal(ciphertext, NONCE_LENGTH, inputLength, plainText, 0);
+      }  catch (AEADBadTagException e) {
+        throw new RuntimeException("GCM tag check failed", e);
```

Comment:
Can we add more information in this exception for debugging friendly? We had several issues earlier with this exception and felt hard to know what is going on.

Reply:
agreed, this one is tricky. I'll add the info.